File: examples/canonical-faker/go/internal/distributions/distributions.go
package distributions

import (
	"math"
	"math/rand"
)

func WeightedPlan(rng *rand.Rand) string {
	return weightedPick(rng, []string{"free", "pro", "enterprise"}, []int{50, 35, 15})
}

func WeightedRegion(rng *rand.Rand) string {
	return weightedPick(rng, []string{"us-east", "us-west", "eu-west", "ap-southeast"}, []int{40, 25, 20, 15})
}

func WeightedLocale(rng *rand.Rand) string {
	return weightedPick(rng, []string{"en-US", "en-GB", "de-DE", "fr-FR"}, []int{60, 20, 10, 10})
}

func WeightedMembershipRole(rng *rand.Rand) string {
	return weightedPick(rng, []string{"owner", "admin", "member", "viewer"}, []int{5, 10, 60, 25})
}

func WeightedProjectStatus(rng *rand.Rand) string {
	return weightedPick(rng, []string{"active", "archived", "draft"}, []int{60, 25, 15})
}

func WeightedAuditAction(rng *rand.Rand) string {
	return weightedPick(rng, []string{"updated", "created", "exported", "invited", "archived", "deleted"}, []int{35, 20, 15, 12, 10, 8})
}

func WeightedInvitationRole(rng *rand.Rand) string {
	return weightedPick(rng, []string{"admin", "member", "viewer"}, []int{15, 65, 20})
}

func MemberCount(rng *rand.Rand, maxUsers int) int {
	return clamp(int(math.Round(rng.NormFloat64()*2+6)), 3, min(50, maxUsers))
}

func ProjectCount(rng *rand.Rand) int {
	return clamp(int(math.Round(rng.NormFloat64()*2.2+3.5)), 1, 20)
}

func APIKeyCount(rng *rand.Rand) int {
	return clamp(int(math.Round(rng.NormFloat64()*2+2)), 0, 10)
}

func InvitationCount(rng *rand.Rand) int {
	return clamp(int(math.Round(rng.NormFloat64()*1.2+1.5)), 0, 5)
}

func AuditEventCount(rng *rand.Rand, status string) int {
	switch status {
	case "active":
		return clamp(int(math.Round(rng.NormFloat64()*5+14)), 8, 30)
	case "archived":
		return clamp(int(math.Round(rng.NormFloat64()*3+8)), 4, 18)
	default:
		return clamp(int(math.Round(rng.NormFloat64()*2+4)), 2, 10)
	}
}

func weightedPick(rng *rand.Rand, values []string, weights []int) string {
	total := 0
	for _, weight := range weights {
		total += weight
	}
	target := rng.Intn(total)
	running := 0
	for index, value := range values {
		running += weights[index]
		if target < running {
			return value
		}
	}
	return values[len(values)-1]
}

func clamp(value int, minimum int, maximum int) int {
	if value < minimum {
		return minimum
	}
	if value > maximum {
		return maximum
	}
	return value
}

func min(left int, right int) int {
	if left < right {
		return left
	}
	return right
}

File: examples/canonical-faker/ruby/lib/distributions.rb
module Distributions
  PLAN_WEIGHTS = { "free" => 50, "pro" => 35, "enterprise" => 15 }.freeze
  REGION_WEIGHTS = { "us-east" => 40, "us-west" => 25, "eu-west" => 20, "ap-southeast" => 15 }.freeze
  LOCALE_WEIGHTS = { "en-US" => 60, "en-GB" => 20, "de-DE" => 10, "fr-FR" => 10 }.freeze
  ROLE_WEIGHTS = { "owner" => 5, "admin" => 10, "member" => 60, "viewer" => 25 }.freeze
  PROJECT_STATUS_WEIGHTS = { "active" => 60, "archived" => 25, "draft" => 15 }.freeze
  AUDIT_ACTION_WEIGHTS = { "updated" => 35, "created" => 20, "exported" => 15, "invited" => 12, "archived" => 10, "deleted" => 8 }.freeze
  RESOURCE_TYPE_WEIGHTS = { "project" => 35, "user" => 15, "membership" => 25, "api_key" => 10, "invitation" => 15 }.freeze
  INVITATION_ROLE_WEIGHTS = { "admin" => 15, "member" => 65, "viewer" => 20 }.freeze

  TIMEZONE_BY_LOCALE = {
    "en-US" => "America/New_York",
    "en-GB" => "Europe/London",
    "de-DE" => "Europe/Berlin",
    "fr-FR" => "Europe/Paris"
  }.freeze

  def self.weighted_choice(weights_hash, rng)
    total = weights_hash.values.sum
    target = rng.rand(total)
    cumulative = 0
    weights_hash.each do |value, weight|
      cumulative += weight
      return value if target < cumulative
    end
  end

  def self.bounded_count(rng, min, max, mode)
    u = rng.rand
    c = (mode - min).to_f / (max - min)
    if u < c
      (min + Math.sqrt(u * (max - min) * (mode - min))).round
    else
      (max - Math.sqrt((1 - u) * (max - min) * (max - mode))).round
    end
  end
end